### ISSUE TITLE
New parser for kotlin leads to performance degradation

### ISSUE THREAD

**Issue by anthony-S93** [fwcd/tree-sitter-kotlin#107]
The problem is described in https://github.com/nvim-treesitter/nvim-treesitter/issues/5657 along with videos demonstrating the issue.

Basically, with the new kotlin parser, there is a noticeable delay in the opening of neovim buffers. The size of the file does not matter. The delay is noticeable even in a small kotlin file with only two lines of code.

**Comment by VladimirMakaev** [comment]
@anthony-S93 Since you did an initial analysis could you pinpoint commit in this repo where regression started?

**Comment by VladimirMakaev** [comment]
Doing the following on latest commit:
```
hyperfine --warmup 3 'tree-sitter parse examples/Logger.kt' -N
Benchmark 1: tree-sitter parse examples/Logger.kt
  Time (mean ± σ):       8.1 ms ±   0.6 ms    [User: 3.4 ms, System: 2.1 ms]
  Range (min … max):     7.4 ms …  11.7 ms    256 runs
```
Please provide more details

**Comment by anthony-S93** [comment]
@VladimirMakaev  Thanks for getting back to me. The issue appeared after updating `nvim-treesitter` from commit `7b26b085` to commit `8996612b`. I've checked the `diff`s in `nvim-treesitter` and compared the revisions for those two commits. At the moment, it seems to me like the corresponding commit that caused the issue is 0ef87892401bb01c84b40916e1f150197bc134b1

The numbers in the benchmark you did seem fine to me. I have no idea why the delay occurred in my case. Though unlikely to be the case (only kotlin files are affected and the issue went away immediately after I removed the kotlin parser entirely), I suppose it could also be an issue with Neovim, though I don't yet have the time to look into it further.

**Comment by lucario387** [comment]
![image](https://github.com/fwcd/tree-sitter-kotlin/assets/29790821/7bf258ee-eb27-46f2-8ffa-c337a2e6e177)
Maybe, just mayyyybe, justtt mayyyyybeeeee, bloating up parser.c by 3 times the size (before was 22.8MB, after that commit it's 70.8MB) is not a good idea

**Comment by fwcd** [comment]
Totally agree, I've reverted the PR in question (#106) and released 0.3.3, which should hopefully fix the issue. This causes too much trouble to be worth it for now and GitHub doesn't like files over 50 MB either.

We can revisit the addition of field names once we find a way to do so at a reasonable parser size.
